File: examples/12_input_rendering.py
"""Example 12: Pluggable Input Rendering — dispatch levels, BAML default, field
flattening, and inline vs template-ref prompts.

Scenario: A produce node receives a Pydantic model with prose fields.
Instead of JSON-dumping it into the prompt (escaped newlines, token waste),
NeoGraph renders it as human-readable text via the renderer dispatch hierarchy:

    Level 1: Model method  — render_for_prompt() on the BaseModel
    Level 2: Node renderer — @node(renderer=XmlRenderer())
    Level 3: Construct/global — Construct(renderer=...) or configure_llm(renderer=...)

Additional sections demonstrate:

    BAML default        — when NO renderer is configured, BaseModel values are
                          rendered via describe_value() in BAML notation.
    Field flattening    — render_for_prompt() returning a BaseModel flattens its
                          fields into individually addressable template vars.
    Inline vs template  — inline prompts (${var.field}) get raw objects with
                          dotted access; template-ref prompts get BAML strings.

This example uses no real LLM. It demonstrates the rendering pipeline
and the render_prompt() / render_input() / build_rendered_input() inspectors.

Run:
    python examples/12_input_rendering.py
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from neograph import (
    Construct,
    Node,
    XmlRenderer,
    DelimitedRenderer,
    JsonRenderer,
    configure_llm,
    describe_type,
    describe_value,
    render_input,
    render_prompt,
)
from neograph.renderers import build_rendered_input


# -- Schemas ----------------------------------------------------------------

class ReadContext(BaseModel, frozen=True):
    """Document content with prose fields."""
    title: str = Field(description="Document title")
    raw: str = Field(description="Raw markdown content")
    tags: list[str] = Field(default_factory=list, description="Classification tags")


class Analysis(BaseModel, frozen=True):
    summary: str
    key_points: list[str]


# -- Level 1: Model with render_for_prompt() method ------------------------

class CustomRendered(BaseModel, frozen=True):
    """Model that controls its own rendering."""
    name: str
    content: str

    def render_for_prompt(self) -> str:
        return f"=== {self.name} ===\n{self.content}"


# -- Fake LLM + prompt compiler (no API keys) ------------------------------

class FakeLLM:
    def with_structured_output(self, model, **kwargs):
        self._model = model
        return self

    def invoke(self, messages, **kwargs):
        return self._model(summary="Example summary", key_points=["point-1"])


def simple_compiler(template, data, **kwargs):
    return [{"role": "user", "content": f"[{template}]\n{data}"}]


configure_llm(
    llm_factory=lambda tier: FakeLLM(),
    prompt_compiler=simple_compiler,
    renderer=XmlRenderer(),  # Level 3: global default
)


# -- Pipeline ---------------------------------------------------------------

analyze = Node(
    "analyze",
    mode="think",
    outputs=Analysis,
    model="fast",
    prompt="analyze",
    renderer=DelimitedRenderer(),  # Level 2: node-level override
)

pipeline = Construct("render-demo", nodes=[analyze])


# -- Run --------------------------------------------------------------------

if __name__ == "__main__":
    doc = ReadContext(
        title="Architecture Decision Record",
        raw="We chose event sourcing for auditability.\nCQRS separates reads from writes.",
        tags=["architecture", "decision"],
    )

    # Level 1: model method wins over any renderer
    custom = CustomRendered(name="spec", content="The system shall log all access.")
    print("Level 1 (model method):")
    print(render_input(custom, renderer=XmlRenderer()))
    print()

    # Level 2: node renderer (DelimitedRenderer on `analyze`)
    print("Level 2 (node renderer) via render_prompt inspector:")
    print(render_prompt(analyze, doc))
    print()

    # Level 3: global renderer (XmlRenderer via configure_llm)
    plain_node = Node("plain", mode="think", outputs=Analysis, model="fast", prompt="p")
    print("Level 3 (global renderer) via render_prompt inspector:")
    print(render_prompt(plain_node, doc))
    print()

    # describe_type for output schema
    print("Output schema (describe_type):")
    print(describe_type(Analysis))
    print()

    # ===================================================================
    # Section: BAML default rendering (no renderer configured)
    # ===================================================================
    # When no renderer is set at any level, BaseModel values are rendered
    # via describe_value() — the BAML notation, not XmlRenderer.

    print("=" * 60)
    print("BAML DEFAULT (no renderer)")
    print("=" * 60)

    # describe_value is the underlying function. render_input dispatches
    # to it when renderer=None and the value is a BaseModel.
    print("\ndescribe_value(doc) — raw BAML notation:")
    print(describe_value(doc))
    print()

    # render_input with renderer=None also produces BAML:
    print("render_input(doc, renderer=None) — same BAML output:")
    print(render_input(doc, renderer=None))
    print()

    # Compare to XmlRenderer to see the difference:
    print("render_input(doc, renderer=XmlRenderer()) — XML for comparison:")
    print(render_input(doc, renderer=XmlRenderer()))
    print()

    # ===================================================================
    # Section: Field flattening via render_for_prompt() returning BaseModel
    # ===================================================================
    # When render_for_prompt() returns a BaseModel (not a str), its fields
    # become individually addressable template vars in template-ref prompts.

    class SourceDoc(BaseModel, frozen=True):
        """The flattened model returned by render_for_prompt."""
        headline: str
        body: str
        provenance: str

    class ResearchReport(BaseModel, frozen=True):
        """Model that flattens itself for prompt consumption."""
        title: str
        raw_content: str
        source_url: str

        def render_for_prompt(self) -> SourceDoc:
            """Return a BaseModel whose fields become separate template vars."""
            return SourceDoc(
                headline=self.title.upper(),
                body=self.raw_content,
                provenance=f"Source: {self.source_url}",
            )

    report = ResearchReport(
        title="Quarterly Review",
        raw_content="Revenue grew 12% year-over-year.",
        source_url="https://example.com/q4",
    )

    print("=" * 60)
    print("FIELD FLATTENING (render_for_prompt returns BaseModel)")
    print("=" * 60)

    # build_rendered_input shows the full RenderedInput structure:
    ri = build_rendered_input({"report": report}, renderer=None)
    print("\nRenderedInput.rendered (whole model, BAML):")
    print(ri.rendered)
    print("\nRenderedInput.flattened (individual fields from SourceDoc):")
    for k, v in ri.flattened.items():
        print(f"  {k}: {v!r}")
    print("\nRenderedInput.for_template_ref (merged — what the prompt compiler sees):")
    tmpl_data = ri.for_template_ref
    for k, v in tmpl_data.items():
        print(f"  {k}: {v!r}")
    print("\navailable_keys_template:", sorted(ri.available_keys_template))
    print("available_keys_inline:  ", sorted(ri.available_keys_inline))
    print()

    # A template-ref prompt can now use {headline}, {body}, {provenance}
    # as first-class vars, not just {report}.

    # ===================================================================
    # Section: Inline prompt vs template-ref prompt
    # ===================================================================
    # Inline prompts contain ${var.field} — they receive raw objects so
    # dotted attribute access works. Template-ref prompts (no spaces, no
    # ${...}) receive BAML-rendered strings.

    print("=" * 60)
    print("INLINE vs TEMPLATE-REF prompt rendering")
    print("=" * 60)

    # Use a node with NO renderer (renderer=None) so BAML is the default.
    # Since the global configure_llm set XmlRenderer, we create a node
    # with explicit renderer=None to show BAML behavior.
    # However, render_prompt uses: effective_renderer = node.renderer or global
    # So we demonstrate inline vs template-ref via build_rendered_input directly.

    print("\n--- Same data, two prompt styles ---")
    print()

    # Build RenderedInput with no renderer (BAML default):
    ri_baml = build_rendered_input(doc, renderer=None)

    print("RenderedInput.raw (what inline ${var.field} sees):")
    print(f"  type: {type(ri_baml.raw).__name__}")
    print(f"  ri_baml.raw.title  = {ri_baml.raw.title!r}")
    print(f"  ri_baml.raw.tags   = {ri_baml.raw.tags!r}")
    print()

    print("RenderedInput.rendered (what template-ref sees — BAML string):")
    print(ri_baml.rendered)
    print()

    # With XmlRenderer for contrast:
    ri_xml = build_rendered_input(doc, renderer=XmlRenderer())

    print("RenderedInput.rendered with XmlRenderer (template-ref sees XML):")
    print(ri_xml.rendered)
    print()

    # Dict-form (fan-in) example showing both views:
    claims_model = Analysis(summary="Strong growth", key_points=["12% rev", "new market"])
    ri_dict = build_rendered_input({"doc": doc, "prior": claims_model}, renderer=None)

    print("Dict-form fan-in (two upstream values, BAML default):")
    print("  Inline keys (raw objects): ", sorted(ri_dict.available_keys_inline))
    print("  Template keys (rendered):  ", sorted(ri_dict.available_keys_template))
    print()
    print("  ri_dict.raw['doc'].title       =", ri_dict.raw["doc"].title)
    print("  ri_dict.rendered['doc']         =")
    for line in ri_dict.rendered["doc"].splitlines():
        print(f"    {line}")
    print("  ri_dict.rendered['prior']       =")
    for line in ri_dict.rendered["prior"].splitlines():
        print(f"    {line}")

File: docs/design/ts-parity/12_input_rendering.ts
// TS PARITY SKETCH — HYPOTHETICAL, NON-COMPILING.
// Python source: examples/12_input_rendering.py
// Proposed API:  docs/design/typescript-port.md (AD-0 transformer, .pipe(), Zod schema layer)
//
// This example is a RENDERING-pipeline showcase, NOT a DAG/@node showcase. It exercises:
//   - the renderer dispatch hierarchy (Level 1 model-method > Level 2 node > Level 3 global)
//   - BAML default via describe_value()
//   - field flattening when render_for_prompt() returns a BaseModel
//   - inline (${var.field}) vs template-ref ({var}) prompt semantics
//   - the inspectors: render_input(), render_prompt(), build_rendered_input(), describe_type/value
//
// The proposed TS design doc's Feature Parity Matrix marks "Renderers (XML, Delimited, JSON)"
// and "describe_type" as Direct — but says NOTHING about ANY of the five bullets above.
// That silence is where this example lands most of its friction. See the // PARITY: notes.

import { z } from "zod";
import {
  Node,
  Construct,
  XmlRenderer,
  DelimitedRenderer,
  describeType,
  describeValue,
  renderInput,
  renderPrompt,
  buildRenderedInput,
  LlmRuntime,
  // PARITY GAP: none of the following are named anywhere in typescript-port.md.
  // They have to be invented for this example to even be expressible:
  type Renderable,          // interface a schema-bound class implements to carry render_for_prompt
  bindRenderer,             // side-channel: attach a render method to a Zod schema
} from "@neograph/core";

// -- Schemas ----------------------------------------------------------------
// Python: `class ReadContext(BaseModel, frozen=True)` with Field(description=...).
// PARITY (Direct-ish): Pydantic model -> Zod object. frozen=True -> readonly inferred type
// + Object.freeze at the data boundary (Zod does not freeze parsed output). Field(description=)
// -> .describe(); load-bearing because describe_type/XmlRenderer(include_field_info) emit it.

const ReadContext = z.object({
  title: z.string().describe("Document title"),
  raw: z.string().describe("Raw markdown content"),
  tags: z.array(z.string()).default([]).describe("Classification tags"),
});
type ReadContext = Readonly<z.infer<typeof ReadContext>>;

const Analysis = z.object({
  summary: z.string(),
  key_points: z.array(z.string()),
});
type Analysis = Readonly<z.infer<typeof Analysis>>;

// -- Level 1: Model with render_for_prompt() method -------------------------
// Python: a method ON the BaseModel — `CustomRendered(...).render_for_prompt()`.
//
// PARITY: **REDESIGN, verging on BLOCKED.** This is the crux of the example.
// A Zod schema is a validator, not a class; z.infer<T> yields a PLAIN object with no methods.
// So a parsed value can never carry `render_for_prompt`. Two escape hatches, both worse:
//
//   (a) class-based model (zod-class / a custom base) so instances carry methods —
//       but then it is NOT the same object the rest of the pipe validates/serializes,
//       and the checkpointer (msgpack) round-trip drops the prototype/method anyway.
//   (b) a schema->renderer side-channel registry (bindRenderer below) — decouples the
//       render override from the value, so `renderInput(value)` alone (as Python calls it)
//       can no longer find it; you must thread the schema in too.
//
// Python's dispatch is `hasattr(value, "render_for_prompt")` — a per-INSTANCE capability check.
// TS has no per-instance capability on a plain data object. Sketching option (a):

class CustomRendered implements Renderable {
  constructor(readonly name: string, readonly content: string) {}
  static schema = z.object({ name: z.string(), content: z.string() });
  renderForPrompt(): string {
    return `=== ${this.name} ===\n${this.content}`;
  }
}
// PARITY: the design doc's renderer row assumes renderers only ever consume schemas/plain
// values. The model-method LEVEL of the dispatch hierarchy has no mention and no home.

// -- Fake LLM + prompt compiler (no API keys) -------------------------------
// PARITY (Direct): withStructuredOutput exists in LangChain.js; the fake is trivial.
class FakeLLM {
  private model!: z.ZodTypeAny;
  withStructuredOutput(model: z.ZodTypeAny) {
    this.model = model;
    return this;
  }
  invoke(_messages: unknown): Analysis {
    return { summary: "Example summary", key_points: ["point-1"] };
  }
}

// Python: `def simple_compiler(template, data, **kwargs)`.
// PARITY (Direct): plain function; **kwargs -> a trailing `opts?: Record<string, unknown>`.
function simpleCompiler(template: string, data: unknown): Array<{ role: string; content: string }> {
  return [{ role: "user", content: `[${template}]\n${data}` }];
}

// Python: LlmRuntime.build(llm_factory=..., prompt_compiler=..., renderer=XmlRenderer())
// PARITY (Redesign-ish): LlmRuntime is NOT in the parity matrix (only configure_llm() is).
// The Level-3 resolution `node.renderer or runtime.renderer or None` must be specced somewhere;
// here it lives on a runtime object like Python.
const runtime = LlmRuntime.build({
  llmFactory: (_tier: string) => new FakeLLM(),
  promptCompiler: simpleCompiler,
  renderer: new XmlRenderer(), // Level 3: global default
});

// -- Pipeline ---------------------------------------------------------------
// Python: programmatic Node(...) — NOTE this example uses the PROGRAMMATIC surface, not @node,
// so AD-0's transformer is IRRELEVANT here. Node config maps field-for-field.
// PARITY (Direct): this is the "TS-first" surface per AD-2.
const analyze = new Node("analyze", {
  mode: "think",
  outputs: Analysis,
  model: "fast",
  prompt: "analyze",
  renderer: new DelimitedRenderer(), // Level 2: node-level override
});

const pipeline = new Construct("render-demo", { nodes: [analyze] });
void pipeline;

// -- Run --------------------------------------------------------------------
async function main() {
  const doc: ReadContext = ReadContext.parse({
    title: "Architecture Decision Record",
    raw: "We chose event sourcing for auditability.\nCQRS separates reads from writes.",
    tags: ["architecture", "decision"],
  });

  // Level 1: model method wins over any renderer.
  // PARITY (REDESIGN): renderInput must accept a method-carrying instance. With option (a)
  // above, `custom` is a CustomRendered INSTANCE (not a parsed Zod value), so this call site
  // diverges from every OTHER renderInput call (which take plain objects). Heterogeneous input.
  const custom = new CustomRendered("spec", "The system shall log all access.");
  console.log("Level 1 (model method):");
  console.log(renderInput(custom, { renderer: new XmlRenderer() }));
  console.log();

  // Level 2: node renderer (DelimitedRenderer on `analyze`).
  // PARITY (Direct): renderPrompt(node, data, {runtime}) — inspector maps cleanly.
  console.log("Level 2 (node renderer) via renderPrompt inspector:");
  console.log(renderPrompt(analyze, doc, { runtime }));
  console.log();

  // Level 3: global renderer (XmlRenderer via the runtime).
  const plainNode = new Node("plain", { mode: "think", outputs: Analysis, model: "fast", prompt: "p" });
  console.log("Level 3 (global renderer) via renderPrompt inspector:");
  console.log(renderPrompt(plainNode, doc, { runtime }));
  console.log();

  // describe_type for output schema.
  // PARITY (Direct, per doc — "already TS-native notation"). Caveat: describeType(Analysis)
  // takes the SCHEMA, and must read `.describe()` field docs + hoist repeated nested schemas
  // out of Zod's ._def. Doable but the two-pass hoist logic is non-trivial to port.
  console.log("Output schema (describeType):");
  console.log(describeType(Analysis));
  console.log();

  // ===================================================================
  // Section: BAML default rendering (no renderer configured)
  // ===================================================================
  console.log("=".repeat(60));
  console.log("BAML DEFAULT (no renderer)");
  console.log("=".repeat(60));

  // Python: describe_value(doc) — renders an INSTANCE as a typed value literal.
  // PARITY (API GAP / REDESIGN): describe_value's whole job is "value + its TYPE -> literal".
  // In Python the instance IS its type (pydantic carries model_fields). A TS plain object has
  // NO attached schema, so describeValue(doc) alone can't know field types/descriptions.
  // Signature MUST widen to describeValue(value, schema). Every call site below diverges.
  console.log("\ndescribeValue(doc) — raw BAML notation:");
  console.log(describeValue(doc, ReadContext)); // PARITY: extra `ReadContext` arg vs Python.
  console.log();

  console.log("renderInput(doc, {renderer: null}) — same BAML output:");
  console.log(renderInput(doc, { renderer: null, schema: ReadContext })); // PARITY: schema needed.
  console.log();

  console.log("renderInput(doc, {renderer: XmlRenderer}) — XML for comparison:");
  console.log(renderInput(doc, { renderer: new XmlRenderer(), schema: ReadContext }));
  console.log();

  // ===================================================================
  // Section: Field flattening via render_for_prompt() returning BaseModel
  // ===================================================================
  // Python: render_for_prompt() returns a SourceDoc BaseModel, whose fields
  // (headline/body/provenance) become individually addressable template vars.
  //
  // PARITY (REDESIGN, compounds the Level-1 problem): the returned value must ALSO be a
  // field-introspectable model so build_rendered_input can flatten it. In TS the method must
  // return { value, schema } (or a class instance with a static schema) so the framework can
  // read `.shape` to flatten. A bare returned object gives keys but no types.
  const SourceDoc = z.object({
    headline: z.string(),
    body: z.string(),
    provenance: z.string(),
  });

  class ResearchReport implements Renderable {
    constructor(
      readonly title: string,
      readonly raw_content: string,
      readonly source_url: string,
    ) {}
    renderForPrompt(): { value: z.infer<typeof SourceDoc>; schema: typeof SourceDoc } {
      // PARITY: must hand back the schema so the flattener can read field names/types.
      return {
        schema: SourceDoc,
        value: {
          headline: this.title.toUpperCase(),
          body: this.raw_content,
          provenance: `Source: ${this.source_url}`,
        },
      };
    }
  }

  const report = new ResearchReport(
    "Quarterly Review",
    "Revenue grew 12% year-over-year.",
    "https://example.com/q4",
  );

  console.log("=".repeat(60));
  console.log("FIELD FLATTENING (renderForPrompt returns a model)");
  console.log("=".repeat(60));

  // PARITY (Direct-ONCE the above redesign is accepted): RenderedInput itself — raw/rendered/
  // flattened/for_template_ref/available_keys_inline/available_keys_template — is a plain data
  // record and ports 1:1. But the WHOLE abstraction is absent from the parity matrix.
  const ri = buildRenderedInput({ report }, { renderer: null });
  console.log("\nRenderedInput.rendered (whole model, BAML):");
  console.log(ri.rendered);
  console.log("\nRenderedInput.flattened (individual fields from SourceDoc):");
  for (const [k, v] of Object.entries(ri.flattened)) console.log(`  ${k}: ${JSON.stringify(v)}`);
  console.log("\nRenderedInput.forTemplateRef (merged — what the prompt compiler sees):");
  for (const [k, v] of Object.entries(ri.forTemplateRef)) console.log(`  ${k}: ${JSON.stringify(v)}`);
  console.log("\navailableKeysTemplate:", [...ri.availableKeysTemplate].sort());
  console.log("availableKeysInline:  ", [...ri.availableKeysInline].sort());
  console.log();

  // ===================================================================
  // Section: Inline prompt vs template-ref prompt
  // ===================================================================
  console.log("=".repeat(60));
  console.log("INLINE vs TEMPLATE-REF prompt rendering");
  console.log("=".repeat(60));

  // Python: ri_baml.raw is the RAW model — inline ${var.field} does getattr chains on it.
  // PARITY (Direct, actually EASIER): a plain JS object supports `raw.title` natively; no
  // getattr machinery. The inline-vs-template string heuristic (has spaces / has ${...})
  // ports as a plain regex. This half of the section is the smoothest port in the file.
  console.log("\n--- Same data, two prompt styles ---\n");
  const riBaml = buildRenderedInput(doc, { renderer: null, schema: ReadContext });
  console.log("RenderedInput.raw (what inline ${var.field} sees):");
  console.log(`  type: ${typeof riBaml.raw}`);
  console.log(`  riBaml.raw.title = ${JSON.stringify((riBaml.raw as ReadContext).title)}`);
  console.log(`  riBaml.raw.tags  = ${JSON.stringify((riBaml.raw as ReadContext).tags)}`);
  console.log();

  console.log("RenderedInput.rendered (template-ref sees — BAML string):");
  console.log(riBaml.rendered);
  console.log();

  const riXml = buildRenderedInput(doc, { renderer: new XmlRenderer(), schema: ReadContext });
  console.log("RenderedInput.rendered with XmlRenderer (template-ref sees XML):");
  console.log(riXml.rendered);
  console.log();

  // Dict-form (fan-in) — two upstream values, both views.
  // PARITY (Direct on the container; each value STILL needs its schema for BAML). Python
  // infers each value's type from the instance; TS dict-form must carry a parallel schema map.
  const claimsModel: Analysis = { summary: "Strong growth", key_points: ["12% rev", "new market"] };
  const riDict = buildRenderedInput(
    { doc, prior: claimsModel },
    { renderer: null, schemas: { doc: ReadContext, prior: Analysis } }, // PARITY: extra schemas map.
  );
  console.log("Dict-form fan-in (two upstream values, BAML default):");
  console.log("  Inline keys (raw objects): ", [...riDict.availableKeysInline].sort());
  console.log("  Template keys (rendered):  ", [...riDict.availableKeysTemplate].sort());
  console.log();
  console.log("  riDict.raw['doc'].title =", (riDict.raw as Record<string, ReadContext>).doc.title);
}

void main;
